# Import data 

Link from UCI repository. You can find the "Autism Screening Adult Data Set" from the following link: https://archive.ics.uci.edu/ml/datasets/Autism%2BScreening%2BAdult. We will be using this data set.

In [29]:
import pandas as pd
import numpy as np
import matplotlib.pylab as plt

In [42]:
filename = "csv_result-Autism-Adult-Data.csv"

In [43]:
df = pd.read_csv(filename)
df.head(10)

id  A1_Score  A2_Score  A3_Score  A4_Score  A5_Score  A6_Score  A7_Score  \
0   1         1         1         1         1         0         0         1   
1   2         1         1         0         1         0         0         0   
2   3         1         1         0         1         1         0         1   
3   4         1         1         0         1         0         0         1   
4   5         1         0         0         0         0         0         0   
5   6         1         1         1         1         1         0         1   
6   7         0         1         0         0         0         0         0   
7   8         1         1         1         1         0         0         0   
8   9         1         1         0         0         1         0         0   
9  10         1         1         1         1         0         1         1   

   A8_Score  A9_Score  ...  gender       ethnicity jundice austim  \
0         1         0  ...       f  White-European      no     no   
1         1         0  ...       m          Latino      no    yes   
2         1         1  ...       m          Latino     yes    yes   
3         1         0  ...       f  White-European      no    yes   
4         1         0  ...       f               ?      no     no   
5         1         1  ...       m          Others     yes     no   
6         1         0  ...       f           Black      no     no   
7         0         1  ...       m  White-European      no     no   
8         1         1  ...       m  White-European      no     no   
9         1         1  ...       m           Asian     yes    yes   

   contry_of_res used_app_before result     age_desc  \
0  United States              no      6  18 and more   
1         Brazil              no      5  18 and more   
2          Spain              no      8  18 and more   
3  United States              no      6  18 and more   
4          Egypt              no      2  18 and more   
5  United States              no      9  18 and more   
6  United States              no      2  18 and more   
7    New Zealand              no      5  18 and more   
8  United States              no      6  18 and more   
9        Bahamas              no      8  18 and more   

                   relation Class/ASD  
0                      Self        NO  
1                      Self        NO  
2                    Parent       YES  
3                      Self        NO  
4                         ?        NO  
5                      Self       YES  
6                      Self        NO  
7                    Parent        NO  
8                      Self        NO  
9  Health care professional       YES  

[10 rows x 22 columns]

## Identify and handle missing values


## Identify missing values

### Convert "?" to NaN

In this dataset, missing data comes with the question mark "?". We replace "?" with NaN (Not a Number), which is Python's default missing value marker, for reasons of computational speed and convenience.

In [44]:
import numpy as np

# replace "?" to NaN
df.replace("?", np.nan, inplace = True)
df.head(10)

id  A1_Score  A2_Score  A3_Score  A4_Score  A5_Score  A6_Score  A7_Score  \
0   1         1         1         1         1         0         0         1   
1   2         1         1         0         1         0         0         0   
2   3         1         1         0         1         1         0         1   
3   4         1         1         0         1         0         0         1   
4   5         1         0         0         0         0         0         0   
5   6         1         1         1         1         1         0         1   
6   7         0         1         0         0         0         0         0   
7   8         1         1         1         1         0         0         0   
8   9         1         1         0         0         1         0         0   
9  10         1         1         1         1         0         1         1   

   A8_Score  A9_Score  ...  gender       ethnicity jundice austim  \
0         1         0  ...       f  White-European      no     no   
1         1         0  ...       m          Latino      no    yes   
2         1         1  ...       m          Latino     yes    yes   
3         1         0  ...       f  White-European      no    yes   
4         1         0  ...       f             NaN      no     no   
5         1         1  ...       m          Others     yes     no   
6         1         0  ...       f           Black      no     no   
7         0         1  ...       m  White-European      no     no   
8         1         1  ...       m  White-European      no     no   
9         1         1  ...       m           Asian     yes    yes   

   contry_of_res used_app_before result     age_desc  \
0  United States              no      6  18 and more   
1         Brazil              no      5  18 and more   
2          Spain              no      8  18 and more   
3  United States              no      6  18 and more   
4          Egypt              no      2  18 and more   
5  United States              no      9  18 and more   
6  United States              no      2  18 and more   
7    New Zealand              no      5  18 and more   
8  United States              no      6  18 and more   
9        Bahamas              no      8  18 and more   

                   relation Class/ASD  
0                      Self        NO  
1                      Self        NO  
2                    Parent       YES  
3                      Self        NO  
4                       NaN        NO  
5                      Self       YES  
6                      Self        NO  
7                    Parent        NO  
8                      Self        NO  
9  Health care professional       YES  

[10 rows x 22 columns]

In [45]:
missing_data = df.isnull()
missing_data.head()

id  A1_Score  A2_Score  A3_Score  A4_Score  A5_Score  A6_Score  \
0  False     False     False     False     False     False     False   
1  False     False     False     False     False     False     False   
2  False     False     False     False     False     False     False   
3  False     False     False     False     False     False     False   
4  False     False     False     False     False     False     False   

   A7_Score  A8_Score  A9_Score  ...  gender  ethnicity  jundice  austim  \
0     False     False     False  ...   False      False    False   False   
1     False     False     False  ...   False      False    False   False   
2     False     False     False  ...   False      False    False   False   
3     False     False     False  ...   False      False    False   False   
4     False     False     False  ...   False       True    False   False   

   contry_of_res  used_app_before  result  age_desc  relation  Class/ASD  
0          False            False   False     False     False      False  
1          False            False   False     False     False      False  
2          False            False   False     False     False      False  
3          False            False   False     False     False      False  
4          False            False   False     False      True      False  

[5 rows x 22 columns]

### Count missing values in each column

In [46]:
for column in missing_data.columns.values.tolist():
    print(column)
    print (missing_data[column].value_counts())
    print("")    

id
False    704
Name: id, dtype: int64

A1_Score
False    704
Name: A1_Score, dtype: int64

A2_Score
False    704
Name: A2_Score, dtype: int64

A3_Score
False    704
Name: A3_Score, dtype: int64

A4_Score
False    704
Name: A4_Score, dtype: int64

A5_Score
False    704
Name: A5_Score, dtype: int64

A6_Score
False    704
Name: A6_Score, dtype: int64

A7_Score
False    704
Name: A7_Score, dtype: int64

A8_Score
False    704
Name: A8_Score, dtype: int64

A9_Score
False    704
Name: A9_Score, dtype: int64

A10_Score
False    704
Name: A10_Score, dtype: int64

age
False    702
True       2
Name: age, dtype: int64

gender
False    704
Name: gender, dtype: int64

ethnicity
False    609
True      95
Name: ethnicity, dtype: int64

jundice
False    704
Name: jundice, dtype: int64

austim
False    704
Name: austim, dtype: int64

contry_of_res
False    704
Name: contry_of_res, dtype: int64

used_app_before
False    704
Name: used_app_before, dtype: int64

result
False    704
Name: result, dtype: i

### Replace missing values

Based on the summary above, each column has 704 rows of data, 3 columns containing missing data:

Relation: 95 missing data

age: 2 missing data

Ethnicity: 95 missing data

Since they are categorical we replace them with the most occuring value in the column.

In [47]:
df = df.fillna(df.mode().iloc[0])
df.head()

id  A1_Score  A2_Score  A3_Score  A4_Score  A5_Score  A6_Score  A7_Score  \
0   1         1         1         1         1         0         0         1   
1   2         1         1         0         1         0         0         0   
2   3         1         1         0         1         1         0         1   
3   4         1         1         0         1         0         0         1   
4   5         1         0         0         0         0         0         0   

   A8_Score  A9_Score  ...  gender       ethnicity jundice austim  \
0         1         0  ...       f  White-European      no     no   
1         1         0  ...       m          Latino      no    yes   
2         1         1  ...       m          Latino     yes    yes   
3         1         0  ...       f  White-European      no    yes   
4         1         0  ...       f  White-European      no     no   

   contry_of_res used_app_before result     age_desc  relation Class/ASD  
0  United States              no      6  18 and more      Self        NO  
1         Brazil              no      5  18 and more      Self        NO  
2          Spain              no      8  18 and more    Parent       YES  
3  United States              no      6  18 and more      Self        NO  
4          Egypt              no      2  18 and more      Self        NO  

[5 rows x 22 columns]

## Removing unnecessary attributes

The attribute named “Result” is the sum total of the scores of questions (A1-A10). The value of “age_desc” is ’18 and more’ in all the samples. These two attributes are unnecessary. 

In [48]:
df = df.drop(['result','age_desc'], axis = 1)
df.head()

id  A1_Score  A2_Score  A3_Score  A4_Score  A5_Score  A6_Score  A7_Score  \
0   1         1         1         1         1         0         0         1   
1   2         1         1         0         1         0         0         0   
2   3         1         1         0         1         1         0         1   
3   4         1         1         0         1         0         0         1   
4   5         1         0         0         0         0         0         0   

   A8_Score  A9_Score  A10_Score age gender       ethnicity jundice austim  \
0         1         0          0  26      f  White-European      no     no   
1         1         0          1  24      m          Latino      no    yes   
2         1         1          1  27      m          Latino     yes    yes   
3         1         0          1  35      f  White-European      no    yes   
4         1         0          0  40      f  White-European      no     no   

   contry_of_res used_app_before relation Class/ASD  
0  United States              no     Self        NO  
1         Brazil              no     Self        NO  
2          Spain              no   Parent       YES  
3  United States              no     Self        NO  
4          Egypt              no     Self        NO


## Rename column names

In [81]:
df.rename(columns={'jundice':'jaundice'},inplace = True)
df.rename(columns={'contry_of_res':'country_of_res'},inplace = True)
df.rename(columns={'austim':'autism'},inplace = True)
df

id  A1_Score  A2_Score  A3_Score  A4_Score  A5_Score  A6_Score  \
0      1         1         1         1         1         0         0   
1      2         1         1         0         1         0         0   
2      3         1         1         0         1         1         0   
3      4         1         1         0         1         0         0   
4      5         1         0         0         0         0         0   
..   ...       ...       ...       ...       ...       ...       ...   
699  700         0         1         0         1         1         0   
700  701         1         0         0         0         0         0   
701  702         1         0         1         1         1         0   
702  703         1         0         0         1         1         0   
703  704         1         0         1         1         1         0   

     A7_Score  A8_Score  A9_Score  A10_Score age gender       ethnicity  \
0           1         1         0          0  26      f  White-European   
1           0         1         0          1  24      m          Latino   
2           1         1         1          1  27      m          Latino   
3           1         1         0          1  35      f  White-European   
4           0         1         0          0  40      f  White-European   
..        ...       ...       ...        ...  ..    ...             ...   
699         1         1         1          1  25      f  White-European   
700         0         1         0          1  34      m        Hispanic   
701         1         1         0          1  24      f  White-European   
702         1         0         1          1  35      m     South Asian   
703         1         1         1          1  26      f  White-European   

    jaundice autism country_of_res used_app_before relation Class/ASD  
0         no     no  United States              no     Self        NO  
1         no    yes         Brazil              no     Self        NO  
2        yes    yes          Spain              no   Parent       YES  
3         no    yes  United States              no     Self        NO  
4         no     no          Egypt              no     Self        NO  
..       ...    ...            ...             ...      ...       ...  
699       no     no         Russia              no     Self       YES  
700       no     no         Mexico              no   Parent        NO  
701       no     no         Russia              no     Self       YES  
702       no     no       Pakistan              no     Self        NO  
703       no     no         Cyprus              no     Self       YES  

[704 rows x 20 columns]

### Removing non numeric data

In [89]:
df.drop_duplicates(inplace = True)
df_num = df.drop(['gender','ethnicity','jaundice','autism','country_of_res','used_app_before','relation','Class/ASD'], axis = 'columns')
df_num

id  A1_Score  A2_Score  A3_Score  A4_Score  A5_Score  A6_Score  \
0      1         1         1         1         1         0         0   
1      2         1         1         0         1         0         0   
2      3         1         1         0         1         1         0   
3      4         1         1         0         1         0         0   
4      5         1         0         0         0         0         0   
..   ...       ...       ...       ...       ...       ...       ...   
699  700         0         1         0         1         1         0   
700  701         1         0         0         0         0         0   
701  702         1         0         1         1         1         0   
702  703         1         0         0         1         1         0   
703  704         1         0         1         1         1         0   

     A7_Score  A8_Score  A9_Score  A10_Score age  
0           1         1         0          0  26  
1           0         1         0          1  24  
2           1         1         1          1  27  
3           1         1         0          1  35  
4           0         1         0          0  40  
..        ...       ...       ...        ...  ..  
699         1         1         1          1  25  
700         0         1         0          1  34  
701         1         1         0          1  24  
702         1         0         1          1  35  
703         1         1         1          1  26  

[704 rows x 12 columns]

## Encoding categorical data.


Relation, Jaundice, Gender, Country_of_res, Ethnicity, Autism , used_app_before, attributes have categorical data.

One hot encoding is used for this purpose since they are nominal variables.

Nominal variable - comprises a finite set of discrete values with no relationship between values.
Ordinal Variable - Variable comprises a finite set of discrete values with a ranked ordering between values.

One hot endcoding is used because in most of the models numeric data is required.


In [83]:
from sklearn.preprocessing import OneHotEncoder
import category_encoders as ce

In [97]:
subset =  df[['jaundice','gender','country_of_res','ethnicity','autism','used_app_before','relation','Class/ASD']]
ce_one_hot = ce.OneHotEncoder(cols = ['jaundice','gender','country_of_res','ethnicity','autism','used_app_before','relation','Class/ASD'])
autism_onehot=ce_one_hot.fit_transform(subset)
autism_onehot

jaundice_1  jaundice_2  gender_1  gender_2  country_of_res_1  \
0             1           0         1         0                 1   
1             1           0         0         1                 0   
2             0           1         0         1                 0   
3             1           0         1         0                 1   
4             1           0         1         0                 0   
..          ...         ...       ...       ...               ...   
699           1           0         1         0                 0   
700           1           0         0         1                 0   
701           1           0         1         0                 0   
702           1           0         0         1                 0   
703           1           0         1         0                 0   

     country_of_res_2  country_of_res_3  country_of_res_4  country_of_res_5  \
0                   0                 0                 0                 0   
1                   1                 0                 0                 0   
2                   0                 1                 0                 0   
3                   0                 0                 0                 0   
4                   0                 0                 1                 0   
..                ...               ...               ...               ...   
699                 0                 0                 0                 0   
700                 0                 0                 0                 0   
701                 0                 0                 0                 0   
702                 0                 0                 0                 0   
703                 0                 0                 0                 0   

     country_of_res_6  ...  autism_2  used_app_before_1  used_app_before_2  \
0                   0  ...         0                  1                  0   
1                   0  ...         1                  1                  0   
2                   0  ...         1                  1                  0   
3                   0  ...         1                  1                  0   
4                   0  ...         0                  1                  0   
..                ...  ...       ...                ...                ...   
699                 0  ...         0                  1                  0   
700                 0  ...         0                  1                  0   
701                 0  ...         0                  1                  0   
702                 0  ...         0                  1                  0   
703                 0  ...         0                  1                  0   

     relation_1  relation_2  relation_3  relation_4  relation_5  Class/ASD_1  \
0             1           0           0           0           0            1   
1             1           0           0           0           0            1   
2             0           1           0           0           0            0   
3             1           0           0           0           0            1   
4             1           0           0           0           0            1   
..          ...         ...         ...         ...         ...          ...   
699           1           0           0           0           0            0   
700           0           1           0           0           0            1   
701           1           0           0           0           0            0   
702           1           0           0           0           0            1   
703           1           0           0           0           0            0   

     Class/ASD_2  
0              0  
1              0  
2              1  
3              0  
4              0  
..           ...  
699            1  
700            0  
701            1  
702            0  
703            1  

[704 rows x 93 columns]

### Creating target attribute (numerical)

In [98]:
final_df = pd.concat([df_num, autism_onehot], axis='columns')
target = final_df[['Class/ASD_1']]
final_df.drop(['Class/ASD_1','Class/ASD_2'],axis = "columns", inplace = True)
final_df.head()

id  A1_Score  A2_Score  A3_Score  A4_Score  A5_Score  A6_Score  A7_Score  \
0   1         1         1         1         1         0         0         1   
1   2         1         1         0         1         0         0         0   
2   3         1         1         0         1         1         0         1   
3   4         1         1         0         1         0         0         1   
4   5         1         0         0         0         0         0         0   

   A8_Score  A9_Score  ...  ethnicity_11 autism_1  autism_2  \
0         1         0  ...             0        1         0   
1         1         0  ...             0        0         1   
2         1         1  ...             0        0         1   
3         1         0  ...             0        0         1   
4         1         0  ...             0        1         0   

   used_app_before_1  used_app_before_2  relation_1  relation_2  relation_3  \
0                  1                  0           1           0           0   
1                  1                  0           1           0           0   
2                  1                  0           0           1           0   
3                  1                  0           1           0           0   
4                  1                  0           1           0           0   

   relation_4  relation_5  
0           0           0  
1           0           0  
2           0           0  
3           0           0  
4           0           0  

[5 rows x 103 columns]

###  Remove recurrent values

In [99]:
final_df = final_df.drop(['jaundice_2','gender_2','autism_2','used_app_before_2'],axis = 'columns')
final_df

id  A1_Score  A2_Score  A3_Score  A4_Score  A5_Score  A6_Score  \
0      1         1         1         1         1         0         0   
1      2         1         1         0         1         0         0   
2      3         1         1         0         1         1         0   
3      4         1         1         0         1         0         0   
4      5         1         0         0         0         0         0   
..   ...       ...       ...       ...       ...       ...       ...   
699  700         0         1         0         1         1         0   
700  701         1         0         0         0         0         0   
701  702         1         0         1         1         1         0   
702  703         1         0         0         1         1         0   
703  704         1         0         1         1         1         0   

     A7_Score  A8_Score  A9_Score  ...  ethnicity_9 ethnicity_10  \
0           1         1         0  ...            0            0   
1           0         1         0  ...            0            0   
2           1         1         1  ...            0            0   
3           1         1         0  ...            0            0   
4           0         1         0  ...            0            0   
..        ...       ...       ...  ...          ...          ...   
699         1         1         1  ...            0            0   
700         0         1         0  ...            1            0   
701         1         1         0  ...            0            0   
702         1         0         1  ...            0            0   
703         1         1         1  ...            0            0   

     ethnicity_11  autism_1  used_app_before_1  relation_1  relation_2  \
0               0         1                  1           1           0   
1               0         0                  1           1           0   
2               0         0                  1           0           1   
3               0         0                  1           1           0   
4               0         1                  1           1           0   
..            ...       ...                ...         ...         ...   
699             0         1                  1           1           0   
700             0         1                  1           0           1   
701             0         1                  1           1           0   
702             0         1                  1           1           0   
703             0         1                  1           1           0   

     relation_3  relation_4  relation_5  
0             0           0           0  
1             0           0           0  
2             0           0           0  
3             0           0           0  
4             0           0           0  
..          ...         ...         ...  
699           0           0           0  
700           0           0           0  
701           0           0           0  
702           0           0           0  
703           0           0           0  

[704 rows x 99 columns]








# Random forest

In [135]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix


x_train, x_test, y_train, y_test = train_test_split(final_df,target,test_size=0.2)

In [136]:
len(y_train)

563

In [137]:
from sklearn.ensemble import RandomForestClassifier
model = RandomForestClassifier(n_estimators = 100)
model.fit(x_train,y_train)

<ipython-input-137-f37a0a60595e>:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  model.fit(x_train,y_train)


RandomForestClassifier()

In [138]:
model.score(x_test, y_test)

0.9645390070921985

In [140]:
y_predicted = model.predict(x_test)

## Evaluatig confusion matrix

In [143]:
cm = confusion_matrix(y_test,y_predicted)
cm

array([[ 35,   4],
       [  1, 101]], dtype=int64)

In [144]:
x_predicted = model.predict(x_train)
cm1 = confusion_matrix(y_train,x_predicted)

In [145]:
cm1

array([[150,   0],
       [  0, 413]], dtype=int64)